### ISSUE TITLE
Add ignore: ["default-namespace"] and ignoreNamespaces: [] options to selector-type-no-unknown

### ISSUE THREAD

**Issue by bedney** [stylelint/stylelint#2326]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

The `selector-type-no-unknown` rule doesn't take CSS namespace into account.

> Which rule, if any, is this issue related to?

`selector-type-no-unknown`

> What CSS is needed to reproduce this issue?

e.g.

```css

@namespace todomvc url(urn:app:todomvc);

todomvc|app {
    display: block;
    position: absolute;
    height: 100%;
    width: 100%;
    margin: 0;
    padding: 0;
}
```

> What stylelint configuration is needed to reproduce this issue?

e.g.

```json
{
    "extends": "stylelint-config-standard",
    "rules": {
        "selector-type-no-unknown": [true, { "severity": "warning" }]
    }
}
```

> Which version of stylelint are you using?

`7.8.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint **/*.css`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No. CSS namespaces are a W3C-approved syntax.

> What did you expect to happen?

"No warnings to be flagged."

> What actually happened (e.g. what warnings or errors you are getting)?

"The following warnings were flagged:"

```shell
   12:8    warn    Unexpected unknown type selector "app"                         selector-type-no-unknown
```

Note that this bug can be fixed by altering `selector-type-no-unknown/index.js` by inserting this at line 85:

```js
if (optionsMatches(options, "ignoreTypes", tagNode.namespace + '|' + tagNode.value)) {
            return
          }
```

and then by updating the `.stylelintrc` to have:

```json
        "selector-type-no-unknown": [true, { 
            "severity": "warning", 
            "ignoreTypes": ["/^.*\\|.*$/"] 
        }],
```

**Comment by hudochenkov** [comment]
Thank you for the report and for using the template.

Rule is working as expected. `todomvc|app` is an `app` type selector in `todomvc` namespace. `app` a is non-standard type, that's why rule shows warning.

Use this config to add `app` as recognised type selector:

```json
{
    "rules": {
        "selector-type-no-unknown": [
            true, 
            {
                "ignoreTypes": ["app"]
             }
        ]
    }
}
```

**Comment by jeddy3** [comment]
> Rule is working as expected.

I agree. 

CSS Namespaces are new to me, however. I'd like to know a little more about how they are typically used in projects, as there _might_ be the case for adding an option here (e.g. `ignore: ["namespaced"]` or changing the behaviour of the rule to ignore namespaced type selectors by default. 

@bedney Can you please tell us a little more about how you use CSS namespaces? For example, do you only namespace custom type selectors or is it a mix of standard and custom?

**Comment by idearat** [comment]
I disagree. If you're going to hard-code support for only html and svg then at a minimum the rule should check tagNode.namespace and ignore those which aren't html or svg. If you're using a different xml namespace, say XMPP for example, it's not practical to expect someone to put every individual tag name in the XMPP namespace into an ignoreTypes rule.

**Comment by idearat** [comment]
To expand on "how you use CSS namespaces" I use them in accordance with the CSS specification. I use an @namespace to declare them and then use CSS to style XML. XML like Atom, XMPP, XML/EDI, XBRL and other more commonly standards as well as enterprise-specific tag sets. 

Styling the original XML directly eliminates the necessity of running a templating engine or templating pipeline simply to transform one form of XML into another form of XML (XHTML) -- or worse to convert XML into JSON and then back into XHTML -- just to end up styling it with the same CSS that would work on the original XML. It's about efficiency in the end.

**Comment by bedney** [comment]
I would agree with @idearat here - there can't be an expectation that I would be able to put an indeterminate set of custom element names in the ignoreTypes rule config.

I use both non-namespaced XHTML and SVG selectors (e.g. `div {...}`) and namespaced ones (e.g. `html|div {...}`) depending on context.

Note the `html` namespace prefix is not hardcoded here (per XML namespace rules). As @idearat points out, it's mapped via the `@namespace` rule (e.g. `@namespace todomvc url(urn:app:todomvc);`). I could map the namespace prefix `fluffy` to the XHTML namespace via this declaration: `@namespace fluffy url(http://www.w3.org/1999/xhtml);` and then use `fluffy|div {...}`.

One last note. It's possible to set the default namespace by using `@namespace` with no prefix. If I declared `@namespace url(urn:app:todomvc);` (note the missing prefix), I could then just use `app {...}`. Of course, at that point, I'd have to map XHTML to a prefix (e.g. `html`) because I'm using the `urn:app:todomvc` namespace as the default namespace. Usually, the browsers set the XHTML namespace to be the default.

I'm not sure how much of this affects the discussion here, but you said you wanted more information about how CSS @namespaces work, so there it is :-).

**Comment by idearat** [comment]
On a somewhat related note, if the goal of this rule is to verify HTML5 compliance it should include support/checks for the MathML tags which are part of the tag set.

**Comment by jeddy3** [comment]
@bedney & @idearat Thanks for taking the time out to explain CSS namespaces. The different use cases detailed [in this comment](https://github.com/stylelint/stylelint/issues/2326#issuecomment-278390031) were particularly informative.

I think we should definitely address the issue. I'm just wondering what is the best approach. We need a solution that takes into account the following:

> ~1 I use both non-namespaced XHTML and SVG selectors (e.g. div {...}) and namespaced ones (e.g. html|div {...}) depending on context.
> ~2 Note the html namespace prefix is not hardcoded here (per XML namespace rules).
> ~3 It's possible to set the default namespace by using `@namespace` with no prefix

I think we should keep the default behaviour the same as to not complicate the usage of the rule for users who don't use namespaces.

As such, do you think the following two options would cover the use cases listed above?: `ignore["default-namespace"]` and `ignoreNamespaces: []"`

So, all type selectors are checked by default (which is the existing behaviour). This covers the `html|div {...}` and `fluffy|div` cases ("~1" and "~2").

The `ignoreNamespaces` accepts an array, e.g. `["app", "other"]` would ignore `todomvc:app`, `other:type`, and covers the original issue.

The `ignore: ["default-namespace"]` is used to ignore the default namespace, and covers the "~3" case.

Do you think this would suffice? Can you think of another means of addressing those cases?

> if the goal of this rule is to verify HTML5 compliance it should include support/checks for the MathML tags which are part of the tag set.

That's a good point. We're [going to be](https://github.com/stylelint/stylelint/pull/2366) adding a list of MathML tags for custom element support. Therefore, it should be trivial to change the default behaviour of the `selector-type-no-unknown` rule to consider them known. I'll create a separate issue for that.

**Comment by bedney** [comment]
@jeddy3 - Thanks for the reply! It's great to have a project where folks are truly interested in making their software better!

@idearat and I have been collaborating offline on an answer to your proposal above. It looks really good. There are a few notes however:

1. The `ignore["default-namespace"]` flag looks good - no problems there

2. This statement:

The ignoreNamespaces accepts an array, e.g. ["app", "other"] would ignore todomvc:app, other:type, and covers the original issue.

seems slightly incorrect and might be just that you mistyped. Shouldn't that be `["todomvc","other"]`?

3. It would be great if the `ignoreNamespaces` option could (like the `ignoreTypes` option) take a RegExp:

`ignoreNamespaces: ["/regex/", "string"]`

Thanks again!

**Comment by jeddy3** [comment]
> seems slightly incorrect and might be just that you mistyped. Shouldn't that be ["todomvc","other"]?

Yes, it should have been :)

> It would be great if the ignoreNamespaces option could (like the ignoreTypes option) take a RegExp. 

Agreed.

I'm about to dash off on holiday. I'll leave this issue open for discussion while I'm away so that the rest of @stylelint/core can jump in if they've any input on the new options. If there aren't any objections by the time I'm back, I'll label this up then as `type: option` and `status: help wanted`.

**Comment by jeddy3** [comment]
OK, no objections. I'll label this up.

@bedney I'll probably have time to add these options sometime this week, but feel free [to contribute](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#adding-an-option-to-an-existing-rule) them if you fancy it.

**Comment by jeddy3** [comment]
@bedney & @idearat FYI, these options are now available in `7.10.0`.

**Comment by bedney** [comment]
Wow that's great! Thank you very much! We'll give it a whirl.

**Comment by idearat** [comment]
Unfortunately the patch doesn't appear to work correctly. Any use of "ingoreNamespaces", regardless of value, appears to turn off all warnings for all namespaces. 

Assume CSS including:

```

@namespace xctrls url("http://www.technicalpursuit.com/2005/xcontrols");
@namespace pclass url("urn:tibet:pseudoclass");

*[data-theme] {

    /* All themed 'xctrls:tabbar' controls */

    xctrls|tabbar {

        & xctrls|tabitem {

            text-align: center;

            padding: 0.1em 1em;

            margin-left: -1px;

            z-index: 1;

            cursor: pointer;
        }

        & xctrls|tabitem:first-of-type {

            margin-left: 0;
        }
    }

    /* All themed, non-disabled 'xctrls:tabbar' controls */

    xctrls|tabbar:not([pclass|disabled]) {
    }

    /* All themed, non-disabled 'xctrls:tabbar' control labels */

    xctrls|tabbar:not([pclass|disabled]) xctrls|label {
    }
}
```

And assume this .stylelintrc entry:

```
        "selector-type-no-unknown": [ true, {
            "ignoreNamespaces": [ "does-not-matter-due-to-bug" ]
        }],
```

Running stylelint will not produce errors regarding the 'xctrls' namespace selector types for which no matching regex or explicit string has been provided. Instead, all warnings appear to be turned off simply by mentioning the ignoreNamespaces key, regardless of value.

**Comment by jeddy3** [comment]
@ideadapt It's a shame it didn't work out of the box for you.

I'm unable to replicate this at https://stylelint.io/demo/, though.

The following config:

```json
{ 
  "rules": {
    "selector-type-no-unknown": [ true, {
      "ignoreNamespaces": [ "does-not-matter-due-to-bug" ]
    }]
  }
}
```

Gives me:

```shell
8:12 error Unexpected unknown type selector "tabbar" (selector-type-no-unknown)
10:18 error Unexpected unknown type selector "tabitem" (selector-type-no-unknown)
23:18 error Unexpected unknown type selector "tabitem" (selector-type-no-unknown)
31:12 error Unexpected unknown type selector "tabbar" (selector-type-no-unknown)
36:12 error Unexpected unknown type selector "tabbar" (selector-type-no-unknown)
```

Whereas the following gives _no warnings_:

```json
{ 
  "rules": {
    "selector-type-no-unknown": [ true, {
      "ignoreNamespaces": [ "xctrls" ]
    }]
  }
}
```

There is [also a _rejection_ test](https://github.com/stylelint/stylelint/blob/master/lib/rules/selector-type-no-unknown/__tests__/index.js#L217-L222) for this.

Let's get to the bottom of this together. First, can you double check that you're using at least `stylelint@7.10`. If you've installed a global version, make sure that's up to date too:

```shell
./node_modules/.bin/stylelint --version
stylelint --version
```
How are you using stylelint? Via the CLI, Node API or a task runner?

Is your code example and config above the smallest reproducible test case you've tried?

**Comment by idearat** [comment]
@jeddy3 Thanks for the quick sanity check. 

We typically use the API so we can integrate stylelint checks with larger linting process more efficiently. 

It turns out that while I'd updated the global version we use for straight command line use the project I was running my tests in had a leftover 7.7.1 version (although the project's package.json had been updated to specify 7.10.1 or greater). Apparently the npm update hadn't been run so I was unaware of the discrepancy when I checked at the package level.

After removing/updating the local project copy which was being pulled in via require() for API use it looks like this is working as described. Sorry for the confusion.

Thanks again for patching this up.

**Comment by jeddy3** [comment]
> Thanks again for patching this up.

No worries. I'm glad you got it working :)
